x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -161,14 +168,17 @@ public void clusterChanged(ClusterChangedEvent event) {
         final Version mappingVersion = oldestIndexMappingVersion(event.state());
         final ClusterHealthStatus indexStatus = indexMetaData == null ? null :
             new ClusterIndexHealth(indexMetaData, event.state().getRoutingTable().index(indexMetaData.getIndex())).getStatus();
+        // index name non-null iff state recovered
         final String concreteIndexName = indexMetaData == null ? INTERNAL_SECURITY_INDEX : indexMetaData.getIndex().getName();
```

Comment:
Normally, I would've opted for a new state flag `isStateRecovered`, but given the amount of flags already existing I think `concreteIndexName` is a fitting substitute because the index name is also semantically (not only practically) linked to the state recovery status - it's easy to reason that we can't name the security index until the state containing all the index names has been recovered.

Reply:
I would prefer a new flag.
I think it's too easy to miss that the `concreteIndexName` variable has additional, non-obvious semantics, and change the default constructor to assign `INTERNAL_SECURITY_INDEX` to that field.
It's really random that I didn't do that when I introduced the index name field.

As the code stands, you hanging this protection off the fact that we
1. happen to set that variable to `null` at construction, 
2. don't (currently) assign a value to it until we get past the `STATE_NOT_RECOVERED_BLOCK`
3. always set it to a non-null value once we've recovered.

I don't think you can be sure that we will never change this implementation to behave differently and break one of those pre-conditions.
If we're worried about the number of fields here, we can switch all the booleans to a `BitSet`, but I don't think it's actually necessary to re-use fields for a purpose other than what they were intended.